fastly/base_fastly_service_v1.go

```diff
@@ -0,0 +1,482 @@
+	if d.HasChange("name") || d.HasChange("comment") {
+		_, err := conn.UpdateService(&gofastly.UpdateServiceInput{
+			ID:      d.Id(),
+			Name:    d.Get("name").(string),
+			Comment: d.Get("comment").(string),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Once activated, Versions are locked and become immutable. This is true for
+	// versions that are no longer active. For Domains, Backends, DefaultHost and
+	// DefaultTTL, a new Version must be created first, and updates posted to that
+	// Version. Loop these attributes and determine if we need to create a new version first
+	var needsChange bool
+	for _, a := range serviceDef.GetAttributeHandler() {
+		if a.HasChange(d) {
+			needsChange = true
+			break
+		}
+	}
+
+	// Update the active version's comment. No new version is required for this
+	if d.HasChange("version_comment") && !needsChange {
+		latestVersion := d.Get("active_version").(int)
+		if latestVersion == 0 {
+			// If the service was just created, there is an empty Version 1 available
+			// that is unlocked and can be updated
+			latestVersion = 1
+		}
+
+		opts := gofastly.UpdateVersionInput{
+			Service: d.Id(),
+			Version: latestVersion,
+			Comment: d.Get("version_comment").(string),
+		}
+
+		log.Printf("[DEBUG] Update Version opts: %#v", opts)
+		_, err := conn.UpdateVersion(&opts)
+		if err != nil {
+			return err
+		}
+	}
+
+	initialVersion := false
+
+	if needsChange {
+		latestVersion := d.Get("active_version").(int)
+		if latestVersion == 0 {
+			initialVersion = true
+			// If the service was just created, there is an empty Version 1 available
+			// that is unlocked and can be updated
+			latestVersion = 1
+		} else {
+			// Clone the latest version, giving us an unlocked version we can modify
+			log.Printf("[DEBUG] Creating clone of version (%d) for updates", latestVersion)
+			newVersion, err := conn.CloneVersion(&gofastly.CloneVersionInput{
+				Service: d.Id(),
+				Version: latestVersion,
+			})
+			if err != nil {
+				return err
+			}
+
+			// The new version number is named "Number", but it's actually a string
+			latestVersion = newVersion.Number
+			d.Set("cloned_version", latestVersion)
+
+			// New versions are not immediately found in the API, or are not
+			// immediately mutable, so we need to sleep a few and let Fastly ready
+			// itself. Typically, 7 seconds is enough
+			log.Print("[DEBUG] Sleeping 7 seconds to allow Fastly Version to be available")
+			time.Sleep(7 * time.Second)
+
+			// Update the cloned version's comment
+			if d.Get("version_comment").(string) != "" {
+				opts := gofastly.UpdateVersionInput{
+					Service: d.Id(),
+					Version: latestVersion,
+					Comment: d.Get("version_comment").(string),
+				}
+
+				log.Printf("[DEBUG] Update Version opts: %#v", opts)
+				_, err := conn.UpdateVersion(&opts)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		for _, a := range serviceDef.GetAttributeHandler() {
```

Comment:
Just noting that we had the following before:

```golang
		if d.HasChange("default_host") || d.HasChange("default_ttl") || (d.Get("default_ttl") == 0 && initialVersion) {
```

Could `GetAttributeHandler` return a different order of the attributes that introduces a bug? (Thinking out loud).

Reply:
Generally speaking, I don't think order matters, but before, we did check `default_host` and `default_ttl` first and I think it could be possible that these aren't evaludated until later, which could mean that [Fastly Settings aren't updated](https://github.com/terraform-providers/terraform-provider-fastly/blob/master/fastly/resource_fastly_service_v1.go#L306-L310) until after we process something like a logging endpoint, for example. I'm not sure if this matters.

Reply:
Otherwise, this looks like logic parallels what is [currently in `master`](https://github.com/terraform-providers/terraform-provider-fastly/blob/5b471c351ba16454ad3905e4d8f36fce3c29263b/fastly/resource_fastly_user_v1.go).